src/main/kotlin/com/terraformation/backend/nursery/db/BatchStore.kt

```diff
@@ -505,83 +533,56 @@ class BatchStore(
-                    val withdrawalIsNewerThanObservation =
-                        withdrawal.withdrawnDate >= latestObservedDate
-
-                    if (withdrawalIsNewerThanObservation) {
-                      val newGerminatingQuantity =
-                          batch.germinatingQuantity - batchWithdrawal.germinatingQuantityWithdrawn
-                      val newNotReadyQuantity =
-                          batch.notReadyQuantity - batchWithdrawal.notReadyQuantityWithdrawn
-                      val newReadyQuantity =
-                          batch.readyQuantity - batchWithdrawal.readyQuantityWithdrawn
-
-                      if (newGerminatingQuantity < 0 ||
-                          newNotReadyQuantity < 0 ||
-                          newReadyQuantity < 0) {
-                        throw BatchInventoryInsufficientException(batchId)
-                      }
-
-                      // Update the batch's quantities to reflect the withdrawal. If another thread
-                      // has updated the batch since we fetched it above, the version number won't
-                      // match and BatchStaleException will be thrown, which will cause us to try
-                      // again.
-                      updateQuantities(
-                          batchId,
-                          batch.version,
-                          newGerminatingQuantity,
-                          newNotReadyQuantity,
-                          newReadyQuantity,
-                          BatchQuantityHistoryType.Computed,
-                          withdrawalId)
-                    } else {
-                      updateVersionedBatch(batchId, batch.version) {
-                        // Just update the version and modified user/timestamp (need a set() call
-                        // here so the query object is of the correct jOOQ type)
-                        it.set(emptyMap<Any, Any>())
-                      }
+                retryVersionedBatchUpdate(batchId) { batch ->
+                  // Usually we want to subtract the withdrawal amounts from the batch's
+                  // available quantities. However, if the user is entering data about an older
+                  // withdrawal, and they have explicitly edited the available quantities more
+                  // recently than the withdrawal, just record the withdrawal without subtracting
+                  // it from inventory.
+                  //
+                  // We figure this out inside the retry loop because the latest observed time
+                  // might have been updated by whatever operation caused our version number to
+                  // be out of date.
+                  val latestObservedDate =
+                      LocalDate.ofInstant(batch.latestObservedTime, nurseryTimeZone)
+                  val withdrawalIsNewerThanObservation =
+                      withdrawal.withdrawnDate >= latestObservedDate
```

Comment:
should this be just '>' to be considered newer?

Reply:
That won't work in the case where you create a batch on the same day you withdraw from it: the withdrawal won't be treated as newer than the initial observation and the quantities won't be deducted.

Fundamentally the problem is that we're prompting the user for a date and we have no way of telling what time of day the withdrawal happened, so we have to break the tie somewhat arbitrarily.